Meziantou.FluentAssertionsAnalyzers/NunitAssertAnalyzerCodeFixProvider.cs

```diff
@@ -543,6 +546,14 @@ private static async Task<Document> Rewrite(Document document, SyntaxNode nodeTo
                         }
                     }
 
+                    bool IsCollection(ArgumentSyntax argumentSyntax)
+                    {
+                        var argumentTypeSymbol = semanticModel.GetOperation(argumentSyntax.Expression, cancellationToken)?.Type;
+                        if (iEnumerableSymbol == null || argumentTypeSymbol == null || argumentTypeSymbol.Equals(stringSymbol, SymbolEqualityComparer.Default))
+                            return false;
+                        return argumentTypeSymbol.AllInterfaces.Any(i => iEnumerableSymbol.Equals(i, SymbolEqualityComparer.Default));

```

Comment:
I think it doesn't work if `argumentTypeSymbol` is `IEnumerable`.  You can add a test such as

````
Assert.That((IEnumerable)collection, Is.EqualTo(new int[0]))
````

Reply:
I'm afraid that FluentAssertions doesn't support it, feels like an omission:
![image](https://user-images.githubusercontent.com/26791344/202138907-bd35f0b4-bb9b-4ee0-bb67-82146fd397dc.png)


I've added a test for a generic `IEnumerable` though.

Reply:
I don't think the logic is ok. The following class implements `IEnumerable` and will pass the condition, so `IsCollection` returns `true`.
````
class SampleEnumerable : IEnumerable { ... }
````

I think you should update the logic

````
var iEnumerableSymbol = compilation.GetTypeByMetadataName("System.Collections.IEnumerable");
````

````
return argumentTypeSymbol.OriginalDefinition.Equals(iEnumerableSymbol, SymbolEqualityComparer.Default) argumentTypeSymbol.AllInterfaces.Any(i => iEnumerableSymbol.OriginalDefinition.Equals(i, SymbolEqualityComparer.Default));
````

Reply:
Sorry, what logic are your proposing, again? (the example doesn't compile :( ).

That's the test that we'd like to pass in addition to the existing ones, right?
```cs

    [Fact]
    public Task Assert_CustomCollection()
    {
        return Assert(
            """
using System;
using System.Collections;
using NUnit.Framework;

class SampleEnumerable : IEnumerable
{
    public IEnumerator GetEnumerator()
    {
        throw new NotImplementedException();
    }
}

class Test
{
    public void MyTest()
    {
        [|Assert.That(new SampleEnumerable(), Is.EqualTo(new int[0]))|];
    }
}
""",
            """
using System;
using System.Collections;
using FluentAssertions;
using NUnit.Framework;

class SampleEnumerable : IEnumerable
{
    public IEnumerator GetEnumerator()
    {
        throw new NotImplementedException();
    }
}

class Test
{
    public void MyTest()
    {
        new SampleEnumerable().Should().Be(new int[0]);
    }
}
""");
    }
```

Reply:
Sorry if it wasn't clear. Based on your previous comment, the code fixer should not handle `IEnumerable` as a collection as it won't be supported by FluentAssertions. I think your implementation of `IsCollection` doesn't handle it correctly and may consider types that only implements `IEnumerable`.

I don't know if the fixer should rewrite to `Should().Be()` as the assertion is different from the NUnit assertion.

Reply:
What do you think about restricting this to built-in collections only, namely the ones defined in the `System.Collections` and `System.Collections.*` namespaces?

Reply:
I don't think we need to restrict the fix to built-in collections. The fix should work for any collection implementing `IEnumerable<T>`. But it should exclude type that only implements `IEnumerable` as FluentAssertions doesn't provide useful assertions for this type. Or maybe it could rewrite the code as `enumerable.Cast<object>().Should().Equal(...)`.

````c#
    [Fact]
    public Task EnumerableTest()
    {
        return Assert(
$$"""
using System.Collections;
using NUnit.Framework;

class CustomEnumerable : IEnumerable
{
    public IEnumerator GetEnumerator () => throw null;
}

class Test
{
    public void MyTest()
    {
        var collection = new CustomEnumerable();
        [||]Assert.That(collection, Is.EqualTo(new int[0]));
    }
}
""",
$$"""
using System.Collections;
using FluentAssertions;
using NUnit.Framework;

class CustomEnumerable : IEnumerable
{
    public IEnumerator GetEnumerator () => throw null;
}

class Test
{
    public void MyTest()
    {
        var collection = new CustomEnumerable();
        collection.OfType<object>().Should().Equal(new int[0]);
    }
}
""");
    }
````

Reply:
I'm afraid that `collection.OfType<object>().Should().Equal(new int[0]);` is not valid in this context too. :(

I've tried:
```
using System.Collections;
using NUnit.Framework;

namespace asdf;
class CustomEnumerable : IEnumerable
{
    private readonly IEnumerable m_Collection;

    public CustomEnumerable(IEnumerable collection)
    {
        m_Collection = collection;
    }
    public IEnumerator GetEnumerator()
    {
        foreach (var item in m_Collection)
        {
            yield return item;
        }
    }
}

class Test
{
    [Test]
    public void MyTest()
    {
        var collection = new CustomEnumerable(new int[0]);
        Assert.That(collection, Is.EqualTo(new int[0]));
    }

    [Test]
    public void MyTest2()
    {
        var collection = new CustomEnumerable(new int[0]);
        collection.OfType<object>().Should().Equal(new int[0]);
    }

    [Test]
    public void MyTest3()
    {
        var collection = new CustomEnumerable(new int[0]);
        collection.Should().Be(new int[0]);
    }
}
```

with the following result:
![image](https://user-images.githubusercontent.com/26791344/205298082-e04a21aa-a5ff-4c90-9796-92d2d7c8ba8e.png)


I decided to exclude non-generic `IEnumerable` from being supported.
I think it's overall an improvement over `main`, and additional work can be done on a separate PR.

Reply:
> I decided to exclude non-generic IEnumerable from being supported.

I'm fine with this decision

Reply:
Don't forget to mark the PR as Ready for Review when the code is ready for review

Reply:
Yep, totally forgot. Thanks. :)